### PR TITLE
HttpANN algorithm to support language-angostic implementations (Re: Issue #20)

### PR DESCRIPTION
**Pull request by alexklibisz**
Thanks @maumueller, @gosha1128 and others for the fruitful discussion over in #20. I think I've arrived at an implementation that could fit the purpose of language-agnostic (big) ANN.

The HttpANN algorithm is designed to make HTTP calls to a server. The server executes all indexing and querying, thus enabling language-agnostic ANN implementations with minimal overhead. The only requirements for the server are:

1. It should implement the JSON-over-HTTP API documented below (copied from httpann.py). Note that this is a 1:1 copy of the BaseANN Python Class API.
2. It should be able to read the vector dataset in the standard binary format used by this competition.

It could in theory even run remotely, although the intended use-case is that the server runs in the same container.

The overhead for data transfer and serialization is minimal. The server only needs to parse the 10k JSON-encoded query vectors and encode the resulting 10k lists of neighbors.

I also included an example implementation which uses scikit-learn. It's too slow for the large datasets, but it works on the smaller random-xs and random-range-xs. So it should be good enough to demonstrate that this algorithm works.

---

Here is the API that a server must implement:

| Method | Route                | Request Body                                                                                               | Expected Status | Response Body                                                              |
| ------ | -------------------- | ---------------------------------------------------------------------------------------------------------- | --------------- | -------------------------------------------------------------------------- |
| POST   | /init                | dictionary of constructor arguments, e.g., {“metric”: “euclidean”, “dimension”: 99 }                       | 200             | { }                                                                        |
| POST   | /load_index          | { "dataset": <dataset name, e.g. "bigann-10m"> }                                                           | 200             | { "load_index": <Boolean indicating whether the index can be loaded> }     |
| POST   | /set_query_arguments | dictionary of query arguments                                                                              | 200             | { }                                                                        |
| POST   | /query               | { “X”: <query vectors as list of lists of floats>, “k”: <number of neighbors to find and keep in memory> } | 200             | { }                                                                        |
| POST   | /range_query         | { “X”: <query vectors as list of lists of floats>, “radius”: <distance to neighbors> }                     | 200             | { }                                                                        |
| POST   | /get_results         | { }                                                                                                        | 200             | { “get_results”: <neighbors as list of lists of ints> }                    |
| POST   | /get_additional      | { }                                                                                                        | 200             | { “get_additional”: <dictionary of arbitrary additional result metadata> } |
| POST   | /get_range_results   | { }                                                                                                        | 200             | { “get_range_results”: <list of three 1-dimensional lists (lims, I, D)> }  |

### CODE REVIEW

File: algos.yaml [hunk 1]
```diff
@@ -90,6 +110,17 @@ deep-10M:
               "nprobe=2,quantizer_efSearch=8",
               "nprobe=4,quantizer_efSearch=4",
               "nprobe=2,quantizer_efSearch=16"]
+    diskann-t2:
```
Comment (alexklibisz) [inline algos.yaml:113]: There were two sections called "deep-10M" so I just moved the "diskann-t2" section up here to deduplicate the sections.

**Comment by alexklibisz** [comment]
It looks like CI is failing because the fit method is never called. I think I can use the --rebuild flag to force this, but how is it working for the other algos without this flag?

**Review by maumueller — COMMENTED** [review]
This looks very nice, thanks @alexklibisz! Have you been able to measure the overhead vs a naive implementation? 

My only suggestion here is to split up `httpann.py` into `base-http.py` (or whatever you deem to fit) and `http-example-sklearn` to show the interaction between the base http wrapper and the actual implementation that you suggest others to use.

File: benchmark/algorithms/httpann.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        json = self.post("load_index", body, 200)
+        return json["load_index"]
+
+    def query(self, X, k):
+        body = dict(X=[arr.tolist() for arr in X], k=k)
+        self.post("query", body, 200)
+
+    def range_query(self, X, radius):
+        body = dict(X=[arr.tolist() for arr in X], radius=radius)
+        self.post("range_query", body, 200)
```
Comment (maumueller) [inline benchmark/algorithms/httpann.py:70]: I'm wondering what the performance penalty of this will be. I would be happy to 

- change the arguments such that the query vector file is exposed
- or add some kind of `prepare` for providing the query vectors

Comment (alexklibisz) [reply benchmark/algorithms/httpann.py:70]: I've done some local testing, and I think it will be fine since the HTTP overhead and JSON serialization is only incurred one time. It was a bigger deal with ann-benchmarks because that framework required a request and serialization for every query.

**Comment by alexklibisz** [comment]
> My only suggestion here is to split up `httpann.py` into `base-http.py` (or whatever you deem to fit) and `http-example-sklearn.py` to show the interaction between the base http wrapper and the actual implementation that you suggest others to use.

Sounds good. I'll break it into two files.

**Comment by maumueller** [comment]
Great, thanks. Is this ready to be merged?

**Comment by alexklibisz** [comment]
I think so. Please squash if you can, as there are several intermediate/incomplete commits in there.

**Comment by alexklibisz** [comment]
I'll resolve the conflicts and also need to add one bit of documentation.. one moment..

**Comment by maumueller** [comment]
Sorry for not getting back to you early. Shall I squash and merge with main?

**Comment by alexklibisz** [comment]
No problem. Yes, good to go. Thanks!

**Comment by maumueller** [comment]
Looking forward to how this is going to be used. Thanks again.
